src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs

```diff
@@ -580,5 +582,639 @@ public void HangingLambdaParsing_Bug14167()
+                }
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+
+        [Fact]
+        public void TestNullCheckedSingleParamInParens()
+        {
+            UsingDeclaration("Func<int, int> func1 = (x!) => x;");
+            N(SyntaxKind.FieldDeclaration);
+            {
+                N(SyntaxKind.VariableDeclaration);
+                {
+                    N(SyntaxKind.GenericName);
+                    N(SyntaxKind.IdentifierToken, "Func");
+                    N(SyntaxKind.TypeArgumentList);
+                    {
+                        N(SyntaxKind.LessThanToken);
+                        N(SyntaxKind.PredefinedType);
+                        {
+                            N(SyntaxKind.IntKeyword);
+                        }
+                        N(SyntaxKind.CommaToken);
+                        N(SyntaxKind.PredefinedType);
+                        {
+                            N(SyntaxKind.IntKeyword);
+                        }
+                        N(SyntaxKind.GreaterThanToken);
+                    }
+                }
+                N(SyntaxKind.VariableDeclarator);
+                {
+                    N(SyntaxKind.IdentifierToken, "func1");
+                }
+                N(SyntaxKind.EqualsValueClause);
+                {
+                    N(SyntaxKind.EqualsToken);
+                    N(SyntaxKind.ParenthesizedLambdaExpression);
+                    {
+                        N(SyntaxKind.ParameterList);
+                        {
+                            N(SyntaxKind.OpenParenToken);
+                            N(SyntaxKind.Parameter);
+                            {
+                                N(SyntaxKind.IdentifierToken, "x");
+                                N(SyntaxKind.ExclamationToken);
+                            }
+                            N(SyntaxKind.CloseParenToken);
+                        }
+                        N(SyntaxKind.EqualsGreaterThanToken);
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                    }
+                }
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+
+        [Fact]
+        public void TestNullCheckedSingleParamNoSpaces()
+        {
+            UsingDeclaration("Func<int, int> func1 = x!=>x;", expectedErrors: new DiagnosticDescription[]
+            {
+                    // (1,25): error CS8712: Space required between explanation-point and equals-sign here.
+                    // Func<int, int> func1 = x!=>x;
+                    Diagnostic(ErrorCode.ERR_NeedSpaceBetweenExclamationAndEquals, "!=").WithLocation(1, 25),
+                    // (1,27): error CS1525: Invalid expression term '>'
+                    // Func<int, int> func1 = x!=>x;
+                    Diagnostic(ErrorCode.ERR_InvalidExprTerm, ">").WithArguments(">").WithLocation(1, 27)
```

Comment:
odd... why do we et this second error?

Reply:
I presume due to the way the tokens are treated as '!=' + '>'. I'll take a look

Reply:
If you want the errors to be as nice as possible you can try and recover from seeing the stray GreaterThan token in this scenario (maybe by proactively eating it?). This test can also be marked PROTOTYPE to deal with in a future PR if you prefer.

Reply:
Been messing with this for an hour and when I get rid of that error, it messes up the tree and the String equals() call fails. Still working on it, will ask around.

Reply:
see comment here: https://github.com/dotnet/roslyn/pull/35826/files#r288335176